// miranda-bender/webpack.config.js
/** @format */

import path from 'path'
import { glob } from 'glob'
import MiniCssExtractPlugin from 'mini-css-extract-plugin'
import CssMinimizerPlugin from 'css-minimizer-webpack-plugin'
import WebpackObfuscatorPlugin from 'webpack-obfuscator'
import RemoveEmptyScriptsPlugin from 'webpack-remove-empty-scripts'
import fs from 'fs'

const outputDir = path.resolve(process.cwd(), 'dist')
const sourceDir = path.resolve(process.cwd(), 'src')

const getEntries = (extension, isProduction) => {
	const entries = {}
	const files = glob.sync(`./src/${extension}/**/*.${extension}`).filter((file) => !path.basename(file).startsWith('_'))
	if (files.length > 0) {
		for (const file of files) {
			const filename = path.basename(file, path.extname(file))
			const minExtension = isProduction ? '.min' : ''
			entries[`${extension}-${filename}${minExtension}`] = './' + file
		}
	}
	return entries
}

const getGlobalAssetsEntry = () => {
	const entries = {}
	const files = glob.sync(`./assets/**/*`)
	const fileEntries = files.filter((file) => fs.statSync(file).isFile())
	if (fileEntries.length > 0) {
		entries['assets'] = fileEntries.map((str) => './' + str)
	}
	return entries
}

export default (env) => {
	const isProduction = env === 'production' ? true : false
	return {
		mode: isProduction ? 'production' : 'development',
		devtool: isProduction ? false : 'source-map',
		entry: {
			...getEntries('js', isProduction),
			...getEntries('scss', isProduction),
			...getEntries('css', isProduction),
			...getGlobalAssetsEntry(),
		},
		output: {
			path: outputDir,
			clean: true,
		},
		plugins: [new MiniCssExtractPlugin(), new RemoveEmptyScriptsPlugin()],
		optimization: {
			usedExports: true,
			...(isProduction && {
				minimize: true,
				minimizer: [new CssMinimizerPlugin(), new WebpackObfuscatorPlugin({ rotateStringArray: true })],
			}),
		},
		module: {
			rules: [
				{
					test: /\.js$/,
					exclude: /node_modules/,
					use: ['babel-loader'],
				},
				{
					test: /\.scss$/i,
					use: [
						MiniCssExtractPlugin.loader,
						{
							loader: 'css-loader',
							options: {
								sourceMap: !isProduction,
							},
						},
						{ loader: 'sass-loader', options: { sassOptions: { outputStyle: 'expanded' }, sourceMap: !isProduction } },
					],
				},
				{
					test: /\.css$/,
					use: [MiniCssExtractPlugin.loader, { loader: 'css-loader', options: { sourceMap: !isProduction } }],
				},
				{
					test: /\.(png|jpe?g|gif|svg|woff2?|ttf|eot|webp)$/,
					type: 'asset/resource',
					generator: {
						filename: 'assets/[name][ext]',
					},
				},
			],
		},
	}
}
